docs/spec/ibc/queues.md

```diff
@@ -0,0 +1,160 @@
+
+### 3.3   Message Contents
+
+Up to this point, we have focused on the semantics of the message key, and how we can produce a unique identifier for every possible message in every possible connection. The actual data written at the location has been left as an opaque blob, put by providing some structure to the messages, we can enable more functionality.
+
+We define every message in a _send queue_ to consist of a well-known type and opaque data. The IBC protocol relies on the type for routing, and lets the appropriate module process the data as it sees fit. The _receipt queue_ stores if it was an error, an optional error code, and an optional return value. We use the same index as the received message, so that the results of _A:q<sub>B.send</sub>[i]_ are stored at _B:q<sub>A.receipt</sub>[i]_. (read: the message at index _i_ in the _send_ queue for chain B as stored on chain A)
+
+_V<sub>send</sub> = (type, data)_
+
+_V<sub>receipt</sub> = (result, [success|error code])_
+
+### 3.4   Sending a Message
+
+A proper implementation of IBC requires all relevant state to be encapsulated, so that other modules can only interact with it via a fixed API (to be defined in the next sections) rather than directly mutating internal state. This allows the IBC module to provide security guarantees.
+
+Sending an IBC packet involves an application module calling the send method of the IBC module with a packet and a destination chain id.  The IBC module must ensure that the destination chain was already properly registered, and that the calling module has permission to write this packet. If so, the IBC module simply pushes the packet to the tail of the _send queue_, which enables all the proofs described above.
+
+The permissioning of which module can write which packet can be defined per type, so this module can maintain any application-level invariants related to this area. Thus, the "coin" module can maintain the constant supply of tokens, while another module can maintain its own invariants, without IBC messages providing a means to escape their encapsulations. The IBC module must associate every supported message type with a particular handler (_f<sub>type</sub>_) and return an error for unsupported types.
+
+_(IBCsend(D, type, data)_ &#8658; _Success)_
+  &#8658; _push(q<sub>D.send</sub> ,V<sub>send</sub>{type, data})_
+
+We also consider how a given blockchain _A_ is expected to receive the packet from a source chain _S_ with a merkle proof, given the current set of trusted headers for that chain, _T<sub>S</sub>_:
+
+_A:IBCreceive(S, M<sub>k,v,h</sub>)_ &#8658; _match_
+  * _q<sub>S.receipt</sub> =_ &#8709; &#8658; _Error("unregistered sender"),_
+  * _k = (\_, reciept, \_)_ &#8658; _Error("must be a send"),_
+  * _k = (d, \_, \_) and d_ &#8800; _A_ &#8658; _Error("sent to a different chain"),_
+  * _k = (\_, send, i) and head(q<sub>S.receipt</sub>)_ &#8800; _i_ &#8658; _Error("out of order"),_
```

Comment:
Shouldn't we buffer messages if they are received out of order? Or we assume that on sending side, we don't send message with index i+1 before receiving receipt for message i?